### DISCUSSION TITLE
Help me understand the limit on number of serverless function deployments

### DISCUSSION THREAD

**Discussion by simonw** [vercel/vercel#4114] [General] [answered]
I'm still trying to get my head around the new limits (see #4029).

I'll use my newest Vercel-hosted project as an example: https://github.com/simonw/til - which publishes this Vercel-hosted site here: https://til.simonwillison.net/ (see [Using a self-rewriting README powered by GitHub Actions to track TILs](https://simonwillison.net/2020/Apr/20/self-rewriting-readme/)).

I've turned on public source code for that site so you can see how it works: https://til.simonwillison.net/_src

<img width="816" alt="Screen Shot 2020-04-21 at 1 39 25 PM" src="https://user-images.githubusercontent.com/9599/79911754-b8cdf700-83d5-11ea-8335-174e297ffc65.png">

As you can see, all traffic to that site is routed to my `index.py` script. The site content is hosted in a SQLite database file called `til.db`.

Every time I commit to https://github.com/simonw/til a GitHub Action runs which deploys a new copy of the site to Vercel.

With a Pro account I believe I get "640 serverless functions" per month. If I run a deploy that doesn't change the `index.py` file but DOES upload a new copy of the `til.db` file, does that count towards my monthly quota? Or do I only use up one of those 640 when I make a source code change to `index.py`?

How about if I change one of the dependencies in my `requirements.txt` file? Presumably that would cause a recompile of my serverless function and deduct one from my quota?

**Comment by simonw** [comment]
Related question: https://vercel.com/docs/v2/platform/limits#serverless-function-size says "The maximum size for a Serverless Function is 50 MB."

In this example app what counts towards that 50MB? Is it the number of bytes in `index.py`, or the number of bytes in that plus the dependencies in `requirements.txt` - and does it include the number of bytes in the `til.db` binary SQLite file I included in the deployment?

Or does the 50MB apply to the sum of all of the files I uploaded for the deployment - the `.py` files AND any `.html` or `.js` or `.css` files that might get served by the CDN AND the `requirements.txt` and SQLite `.db` files?

**Reply by simonw** [reply]
From https://github.com/zeit/now/discussions/4114#discussioncomment-5935 it sounds like the size limit includes every file I upload, with the exception of any "static files" which can be excluded from the function build. But `@now/python` doesn't exclude anything by default.

**Comment by styfle** [comment, marked as answer]
> If I run a deploy that doesn't change the index.py file but DOES upload a new copy of the til.db file

The `til.db` is bundled in the Serverless Function output so it will be deployed and counted towards your quota.

> How about if I change one of the dependencies in my requirements.txt file?

Yes, for the same reason, these dependencies are installed and then bundled in the Serverless Function output.

> Or does the 50MB apply to the sum of all of the files I uploaded for the deployment 

Yes, its all the files that get bundled. Since `@now/python` isn't using static analysis, its everything. 

See #2830 for a solution using `excludeFiles` pattern to exclude static files from being included in the serverless function output.

**Reply by simonw** [reply]
If I put my `til.db` file in `excludeFiles` will it still be visible on disk to my serverless function running in `index.py` but no longer count as a serverless function deployment?

How about if I deploy `index.db` using excludeFiles (so it's available as a static asset) and then have my serverless function retrieve it via an HTTP request at runtime? Would that allow me to deploy updated binary database files without "spending" a serverless function compilation?

**Reply by simonw** [reply]
The example in #2830 looks like this:
```
{
  "functions": {
    "api/**/*.py": {
      "excludeFiles": "public/**"
    }
  }
}
```
But my `now.json` looks like this - it doesn't use the `"functions"` key at all:
```
{
    "name": "simon-til",
    "version": 2,
    "builds": [
        {
            "src": "index.py",
            "use": "@now/python"
        }
    ],
    "routes": [
        {
            "src": "(.*)",
            "dest": "index.py"
        }
    ]
}
```
How do I rewrite my `now.json` file to use that `"functions"` key and the `"excludeFiles"` option?

**Reply by styfle** [reply]
> If I put my til.db file in excludeFiles will it still be visible on disk to my serverless function running in index.py but no longer count as a serverless function deployment?

No. There is no "disk" so to speak. There are only the files included in the serverless function.

**Reply by styfle** [reply]
> serverless function retrieve it via an HTTP request at runtime?

You could do this but I wouldn't recommend it. As the function scales up and down, the file would need to be retrieved each time. Its not something that would happen once but would happen every time.

At that point, you might as well use a full remote DB instead of sqlite.

**Reply by styfle** [reply]
```json
{
    "builds": [
        {
            "src": "index.py",
            "use": "@now/python",
            "config": { "excludeFiles": "public/**" }
        }
    ],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "/index.py"
        }
    ]
}
```

**Reply by simonw** [reply]
> As the function scales up and down, the file would need to be retrieved each time

Sure, I get that - but the DB file is 76KB (over time I expect it to grow to may one or two MBs) so it's smaller than a jpg or many JSON APIs I might be hitting!

Also I believe lambdas "stay warm" so I could fetch it on cold-start and it would be reused for subsequent requests until the lambda dials down again.

**Reply by styfle** [reply]
> but the DB file is 76KB

That should be fine 👍 

>  lambdas "stay warm" so I could fetch it on cold-start

Indeed, but this is unpredictable because its based on the traffic you receive. If traffic spikes, there will be cold starts as new lambdas are initialized. Similarly, if traffic decreases to 0 then there will be a cold start on the next request.

So you might hit timeouts if you are on the Hobby plan and need to download a 5GB database for each cold start. "Your mileage may vary" so to speak.

**Reply by simonw** [reply]
Yeah downloading a big database would definitely be unpleasant! I think there's a size limit on temporary disk space available to help avoid that though (not sure if that's documented anywhere).

**Comment by ctrlplusb** [comment]
Bit confused about the "serverless function limits per deployment".

Does this mean we have an upper bound of 12 API routes for hobby and 24 API routes for PRO? This seems like a really small number of routes. I'm almost certainly going to have to go ENTERPRISE as I use like 5 different routes purely for auth related stuff (register, login, forgotten password, me).

Does this mean that functionality has to be consolidated so that instead I would have a single "auth" route that would service all of the previous? I feel like this moves away from the essence of serverless.

Guess maybe it is better to shove everything into a single GraphQL EP then. 😅

**Reply by styfle** [reply]
> This seems like a really small number of routes

Remember that this does not include static routes such as `/index.html` or `favicon.ico`. The limit only applies to Serverless Functions. If you need more than 24 Serverless Functions in a single deployment than you will need an Enterprise plan.

**Reply by ctrlplusb** [reply]
Ok, as I was a previous customer I am still on the old unlimited plan. Would there be a cut off point where I would be forced to change to the newer plans?

**Reply by styfle** [reply]
Your plan will remain the same unless you explicitly change it.

**Reply by ctrlplusb** [reply]
Phew, great news. I was getting cold sweats at the idea of a rewrite.

**Reply by ctrlplusb** [reply]
I suspect lots of >= medium size projects will need more that 24 functions though. Perhaps bumping up to ENTERPRISE isn't such a bad thing, but the word itself is off putting. I hear ENTERPRISE I think Microsoft et al size. Wonder if you would consider renaming the plan to make it less associative. Perhaps something simpler like CUSTOM.

**Reply by styfle** [reply]
Yep, Enterprise and Custom are synonymous 👍 

The name "Enterprise" is commonly used to mean you have custom needs that don't fit a particular plan.

![image](https://user-images.githubusercontent.com/229881/80020105-817b4b00-84a6-11ea-9d76-a023ca409382.png)

**Reply by roelvan** [reply]
I just talked to support and I got this reply:
> The enterprise tier starts at four digits of USD. If you want to discuss this, I can redirect this ticket to our sales department.
Best Regards, ~ Paulo De Mitri @ Vercel

This does not feel like "custom needs" but is really enterprise.

And I got this info too:
> Please consider using a different provider for your backend. Vercel excels at delivering an excellent workflow for your frontend and static deployments. Serverless functions should act as helpers, and this limit exists to assist you in that regard.

So how can you build a serious Next.js project with this limitation? Are there any examples out there? What provider do you guys recommend? What works well with Next.js?

**Reply by styfle** [reply]
For Next.js, there is a change coming soon so that so that SSR pages are not impacted by these limits in the same way as Serverless Functions.

See the comment here: https://github.com/zeit/now/discussions/4029#discussioncomment-4455

**Reply by cdock1029** [reply]
@ctrlplusb @roelvan  this might help you re:using fewer functions. Added benefit is you'll more often be hitting a hot function avoiding cold start. Arguably code organization is better as well. As far just "api" functions seems like you could keep it at 1 for most projects using this approach. Then the free/pro tiers don't seem so bad in that context.

https://github.com/zeit/now/discussions/4029#discussioncomment-10322

**Comment by leerob** [comment]
Replaced by https://github.com/vercel/vercel/discussions/6516, please continue the discussion there 🙏
